### ISSUE TITLE
Different font size in paragraphs

### ISSUE THREAD

**Issue by Kerberos3000** [johnfactotum/foliate#288]
Hello, 

I am facing a problem with one specific epub where the font size between paragraphs changes. 
See attached screenshot: 
![Bildschirmfoto vom 2020-04-09 20-38-58](https://user-images.githubusercontent.com/15262355/78929420-aaf9a700-7aa2-11ea-9246-16016e5574bc.png)

With other epub viewer, the font size is consistent and always the same. 

I am using: 
- Foliate 2.0.0
- Linux Mint 19 Cinnamon

The problem seems to be related to this specific epub. I do not face the problem with other epubs. 
(At least I checked two other epubs)

Do you have any idea how to fix this?

**Comment by jkrass1** [comment]
Can confirm
![Screenshot_20200410_154130](https://user-images.githubusercontent.com/9212851/78994613-30d02d80-7b49-11ea-9371-dac81f84751d.png)

**Comment by johnfactotum** [comment]
It's the same issue as https://github.com/johnfactotum/foliate/issues/275, actually.

The problem is that `<a class="test" />` is not valid HTML. If you have:
```html
<!DOCTYPE html>
<meta charset="utf-8">
<style>
a.test {
    font-size: x-small;
}
</style>
<p><a class="test" />Hello</p>
```
This will cause the text to be small as the browser treats the closed anchor as a starting tag instead.

However in XHTML (you have to name it with `.xhtml`, otherwise it gets served as `text/html`), it does work.
```xml
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<html>
    <head>
        <meta charset="utf-8" />
        <style>
        a.test {
            font-size: x-small;
        }
        </style>
    </head>
    <body>
        <p><a class="test" />Hello</p>
    </body>
</html>
```
The text, in this case, is not small.

---

**TL;DR:** the book uses XHTML-isms in `.html` files, so things break. Apart from getting the publisher to fix their books, there's really nothing we can do, except maybe workarounds like:
- edit this book yourself and remove styling on the `.indexterm` class used on the `a` tags; it kind of baffles me why they would add font size styling to a class that is supposed to be empty
- provide a more forceful mode of overriding styles: https://github.com/johnfactotum/foliate/issues/267
- provide a legacy mode that forces serving the page as XHTML, and maybe try to detect if the book is using XHTML (no idea if this would be doable)

**Comment by jkrass1** [comment]
Tbh I don't get it.

In my case, it has some text between <a></a> tags and if you disable the font style for that tag, the text becomes normal

![brave_nsru5AX3QO](https://user-images.githubusercontent.com/9212851/79005390-6f70e280-7b5f-11ea-8660-5dde0d6734d5.jpg)

![brave_INtixNC8rN](https://user-images.githubusercontent.com/9212851/79005392-713aa600-7b5f-11ea-8423-dbfe964f1b0b.jpg)

I mean it looks like valid html for me but still the issue with how the publisher formatted the text.

Overriding would be nice.

**Comment by jkrass1** [comment]
Weird. I checked other ebook reader and the text I mentioned before is outside <a> tags.

![brave_feKMoZbiKK](https://user-images.githubusercontent.com/9212851/79006225-5d903f00-7b61-11ea-9ea3-96541778425a.jpg)

In other part it overrides the font size

![brave_t9DIMRpD0K](https://user-images.githubusercontent.com/9212851/79006293-831d4880-7b61-11ea-9df5-f887404e55d6.jpg)

**Comment by jkrass1** [comment]
> then it normalizes it to `<p><a class="test">Some text</a></p>`

I checked the source code and you are right. 
Thank you for the clarification

**Comment by Kerberos3000** [comment]
Hi all, 

thank you for all the information. 
I did not get everything but I accidentially found out a workaround that at least solved the issue for my epub. ;-) 

- Open the epub in calibre Ebook editor (not Ebook Viewer!)
- Tools -> Repair HTML
- Save 

After that the issue with the different font sizes was solved in Foliate.

@jkrass1 can you confirm this with your epub, too?  

@johnfactotum: Thank you for the support and the great tool! :-)

**Comment by jkrass1** [comment]
@Kerberos3000 it helped!
It adds closing tag for `<a>` so the text appears outside of it.

Looks like other readers do the same https://github.com/johnfactotum/foliate/issues/288#issuecomment-612108668

**Comment by johnfactotum** [comment]
I've found that the mimetype is determined by Epub.js, not by the browser (although the behavior is the same), so forcing it to parse as XHTML is definitely doable. I think we can simply provide an XHML mode; trying to guess the type is a bit too much work, and might be unreliable. But maybe that would be doable, too.

For now the workaround is to apply the following patch:

```diff
diff --git a/src/assets/epub.js b/src/assets/epub.js
index 201281f..d5a3e6d 100644
--- a/src/assets/epub.js
+++ b/src/assets/epub.js
@@ -19200,7 +19200,7 @@ var Archive = function () {
 			} else if (type == "xhtml") {
 				r = (0, _core.parse)(response, "application/xhtml+xml");
 			} else if (type == "html" || type == "htm") {
-				r = (0, _core.parse)(response, "text/html");
+				r = (0, _core.parse)(response, "application/xhtml+xml");
 			} else {
 				r = response;
 			}
```

This will make it parse everything as XHTML, and it does fix this specific issue.

**Comment by Kerberos3000** [comment]
Thank you, the proposed fix worked for me. 
Will you add it in the tool?

**Comment by johnfactotum** [comment]
This fix can't be applied as is, otherwise it would break EPUB 3. I think apart from (or instead of) a manual option, which would probably be confusing and not very useful for most users, perhaps we can implement a simple check for `/<\s*a[^>]*\/>/gi` by default.
